File: utils/bash_container.go
package main

import (
	"os"
	"fmt"

	"flag"
	"math/rand"
	"time"
	"strings"
	"net/http"
	"encoding/json"
	"io/ioutil"

	"github.com/docker/docker/client"
	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/filters"
	"golang.org/x/net/context"
	"io"
)

type appS struct{
	Id 		string	`json:"id"`
	AppId string	`json:"appId"`
	Host 	string	`json:"host"`
	State string	`json:"state"`
}

type taskResponse struct{
	TaskR	[]appS	`json:"tasks"`
}


var Master_Mesos = [3]string{"192.168.131.11","192.168.131.12","192.168.131.13"}

const Version = "1.0"

func main(){
	if len(os.Args) < 1{
		fmt.Println("need service-path")
		os.Exit(-1)
	}

	if os.Args[0] == "-v"{
		fmt.Println("Version",Version)
	}

	var (
		svc string
	)
	flag.StringVar(&svc,"svc","","service path on DC/OS: service.group or group/service")
	flag.Parse()

	svcRunning := getRunningAppbyPath(svc)
	if len(svcRunning) > 1 {
		fmt.Println("this service has many instance. you can use this command to bash")
		for _,s := range svcRunning{
			fmt.Println("DOCKER_HOST=",s.Host,":4243", " docker ps -q --filter \"label=MESOS_TASK_ID=",s.Id,"\"")
			fmt.Println("DOCKER_HOST=",s.Host,":4243", " docker exec -ti <container id> bash")
		}
		os.Exit(-1)
	}
	bashIn(svcRunning)
	fmt.Println(svcRunning)
}

func getAppList(){

}

func bashIn(svcRunning []appS){
	if len(svcRunning) < 1{
		return
	}
	fmt.Println(len(svcRunning))
	for _,s := range svcRunning{
		fmt.Println(s)

		env := envToMap()
		defer mapToEnv(env)

		envMap := map[string]string{
			"DOCKER_HOST":        "tcp://"+s.Host+":4243",
			"DOCKER_API_VERSION": "1.22",
		}
		mapToEnv(envMap)

		dclient, err := client.NewEnvClient()
		if err != nil{
			fmt.Println(err.Error())
			os.Exit(-1)
		}
		//defer dclient.Close()
		filter := filters.NewArgs()
		filter.Add("label", "MESOS_TASK_ID="+s.Id)
		containers, err := dclient.ContainerList(context.Background(), types.ContainerListOptions{
			Size:    true,
			All:     true,
			Latest:	 true,
			Since:   "container",
			Filters: filter})
		if err != nil{
			fmt.Println(err.Error())
			os.Exit(-1)
		}
		for _,c := range containers{

			//execCmd := exec.Command("docker","-H "+s.Host+":4243"," exec "," -ti ",c.ID," bash")
			//execCmd.Stdout = os.Stdout
			//execCmd.Stderr = os.Stderr
			//execCmd.Stdin = os.Stdin
			//err = execCmd.Start()
			//if err !=nil {
			//	fmt.Println(err.Error())
			//}
			//execCmd.Wait()
			//execCmd.Run()
			execId,_ := dclient.ContainerExecCreate(context.Background(),c.ID,types.ExecConfig{

				Tty: true,
				AttachStdin: true,
				AttachStderr: true,
				AttachStdout: true,
				Cmd: []string{"/bin/sh"},
			})
			//err := dclient.ContainerExecStart(context.Background(),execId.ID,types.ExecStartCheck{})
			//if err != nil {
			//	fmt.Println(err)
			//}
			resp,err := dclient.ContainerExecAttach(context.Background(),execId.ID,types.ExecConfig{
				Tty: true,
				AttachStdin: true,
				AttachStderr: true,
				AttachStdout: true,
				Cmd: []string{"/bin/sh"},
				},
			)
			if err != nil {
				return
			}
			defer resp.Close()

			receiveStdout := make(chan error, 1)
			go func() {
				_, err := io.Copy(os.Stdout, resp.Reader)
				receiveStdout <- err
			}()
			select {
			case err := <-receiveStdout:
				if err != nil {
					fmt.Println("Error receiveStdout: %s", err)
					return
				}
			}

			b, err := ioutil.ReadAll(resp.Reader)
			fmt.Printf("%s", b)
		}
	}
}

func getRunningAppbyPath(p string)[]appS{
	if strings.Contains(p,"."){
		fmt.Println("need group/service")
		os.Exit(-1)
	}
	sed := time.Now().Nanosecond()
	master := Master_Mesos[rand.Intn(sed)%3]
	appUrl := strings.Join([]string{master,"marathon","v2/apps",p,"tasks"},"/")

	fmt.Println(appUrl)
	resp, err := http.Get("http://"+appUrl)
	if err != nil{
		fmt.Println("app get error.")
		os.Exit(-1)
	}
	defer resp.Body.Close()

	var data []byte
	data, err = ioutil.ReadAll(resp.Body)
	var appR taskResponse
	err = json.Unmarshal(data,&appR)
	if err != nil{
		fmt.Println("unmarshal error")
	}
	//fmt.Println(string(data))
	//fmt.Println(appR.TaskR)
	return appR.TaskR
}

func mapToEnv(env map[string]string) {
	for k, v := range env {
		os.Setenv(k, v)
	}
}

func envToMap() map[string]string {
	env := make(map[string]string)
	for _, e := range os.Environ() {
		kv := strings.SplitAfterN(e, "=", 2)
		env[kv[0]] = kv[1]
	}

	return env
}

File: utils/bash_container.py
#!/usr/bin/env python

import os
import sys
import time
import json
# import urllib2
import urllib
import random
import subprocess
import docker
import logging
import signal


# type appS struct{
# 	Id 		string	`json:"id"`
# 	AppId string	`json:"appId"`
# 	Host 	string	`json:"host"`
# 	State string	`json:"state"`
# }
#
# type taskResponse struct{
# 	TaskR	[]appS	`json:"tasks"`
# }

OS_ERROR_CODE = 0xff
Master_Mesos = ["192.168.131.11","192.168.131.12","192.168.131.13"]

class AppS(object):

    def __init__(self, id, appId, host, state, startedAt,**kwargs):
        self.Id = id
        self.AppId = appId
        self.Host = host
        self.State = state
        self.StartedAt = startedAt

class TaskR(object):

    def __init__(self, tasks):
        self.tasks = tasks

    def getAppS(self):
        for t in self.tasks:
            yield AppS(**t)

def parseApp(app_path=''):

    if not app_path :
        return False
    elif "." in app_path:
        strs = app_path.split('.')
        strs.reverse()
        return "/".join(strs)
    else:
        return app_path

def update(name):
    url_path = "macos/bash-appM" \
        if name == 'bash-appM' \
        else "ubuntu/bash-appU"
    tmp_path = os.path.join("/tmp",name)
    cmd = "wget http://download.marathon.slave.mesos:31080/utils/%s  -O %s" % (url_path,tmp_path)
    subprocess.call(cmd,shell=True)

    cmd = "chmod +x %s" % (tmp_path)
    subprocess.call(cmd,shell=True)
    pid = os.fork()
    if pid == 0:
        os.rename(tmp_path,"/usr/local/bin/%s"%name)
        print name, "update ok."
    else:
        return

def main():
    if len(sys.argv) < 2:
        print "need app path, Usage: %s path-to-app"%sys.argv[0]
        print "\t Or Usage: %s --update to upgrade."
        sys.exit(-1)

    if sys.argv[1] == "--update":
        update(os.path.basename(sys.argv[0]))
        sys.exit(0)

    app_path = parseApp(sys.argv[1])

    print "App: ", app_path


    master = Master_Mesos[random.randint(0, 2)]

    app_url = "http://" + "/".join([master,"marathon","v2/apps", app_path,"tasks"])
    resp = urllib.urlopen(app_url)

    if resp.code != 200:
        print "Get App tasks error.",resp.read()
        sys.exit(-1)
    data = json.loads(resp.read())
    ts = TaskR(**data)


    if len(ts.tasks) == 0:
        print "\t no container running."

    for apps in ts.getAppS():
        client = docker.DockerClient(base_url=apps.Host+":4243",version="1.24",timeout=10)
        containers_id = client.containers.list(filters={"label":"MESOS_TASK_ID=%s"%apps.Id})


        for id in containers_id:
            print "\tHost:      ", apps.Host
            print "\tState:     ",apps.State
            print "\tStartedAt: ",apps.StartedAt
            print "\tLabel:      label=MESOS_TASK_ID=%s"%apps.Id

            cmd = "docker exec -ti %s bash"%(id.id)
            if subprocesscmd(cmd, env={'DOCKER_HOST': apps.Host+":4243"}) < 0 :
                #use sh
                cmd = 'docker exec -ti %s sh' % (id.id)
                subprocesscmd(cmd, env={'DOCKER_HOST': apps.Host+":4243"})


def subprocesscmd(cmd_str='', timeout=None, description='', env=os.environ,
                  show_message=True):

    os_env = os.environ
    env = os_env.update(env)
    poll_time = 0.2
    _time_begin = time.time()
    if show_message:
        stdout = None
        stderr = None
    else:
        stdout = subprocess.PIPE
        stderr = subprocess.PIPE
    try:
        ret = subprocess.Popen(cmd_str, stdout=stdout, stderr=stderr,
                               shell=True, env=env)
    except OSError as e:
        logging.error('%s %s %s %s' % (description, e, cmd_str, str(env)))
        return -1
    try:
        if timeout:
            deadtime = _time_begin + timeout
            while time.time() < deadtime and ret.poll() is None:
                time.sleep(poll_time)
        else:
            print "In Container "
            ret.wait()
            print "Out Container "

    except KeyboardInterrupt:
        ret.send_signal(signal.SIGINT)
        logging.error('Aborted by user.')
        return -1

    _exec_time = int((time.time() - _time_begin) * 1000)  # ms

    if ret.poll() is None:
        ret.send_signal(signal.SIGINT)
        logging.error(
            '%s : Exec [%s] overtime.' % (description, cmd_str))
        return -_exec_time

    if not show_message:
        for line in ret.stdout:
            if line:
                logging.info('%s %s' % (description, line.strip('\n')))
        for line in ret.stderr:
            if line:
                logging.error('%s %s' % (description, line.strip('\n')))

    if ret.returncode == 0:
        return _exec_time
    else:
        return -_exec_time

if __name__ == '__main__':
    main()